### DISCUSSION TITLE
Reduction in "image quality" while performing colour correction.

### DISCUSSION THREAD

**Discussion by MadhuSN** [colour-science/colour#1137] [Q&A]
I have been working on a project involving colour correction using colour checker card (Spyder checkr 24). 
When I use the colour science library for the colour correction part, there is a significant reduction of quality in the image. 
I have noticed that this happens at the input itself. 
Could anyone help me out here and let me know on how do I retain the quality of image upon colour correction ?
![fabric_3](https://user-images.githubusercontent.com/51255741/232350472-00c3ce1d-16fa-4e94-b7ce-12fdb5b71910.jpg)
![Figure_1_image_after_read](https://user-images.githubusercontent.com/51255741/232350505-94a73099-08fd-41ae-b310-73d0c43bb9d2.png)

**Comment by tjdcs** [comment]
Can you post the before / after examples. "Image quality reduction" is very generic. But if you are thinking about image noise, color correction is known to increase image noise typically.

**Reply by thomasmansencal** [reply]
To extend on what @tjdcs said, the code you used to perform the colour correction would also be useful!

**Reply by MadhuSN** [reply]
> Can you post the before / after examples. "Image quality reduction" is very generic. But if you are thinking about image noise, color correction is known to increase image noise typically.

I have updated the discussion with a sample input picture and a picture of the image after reading the image. Is this happening because of image noise ?

**Reply by MadhuSN** [reply]
> To extend on what @tjdcs said, the code you used to perform the colour correction would also be useful!



  def colour_matching(test_image_1, swatch_values):

    img1 = colour.cctf_decoding(colour.io.read_image(test_image_1))

    colour.plotting.plot_image(img1)

    D65 = colour.CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer']['D65']

    REFERENCE_COLOUR_CHECKER = colour.CCS_COLOURCHECKERS ['babel_average']
    
    REFERENCE_SWATCHES = colour.XYZ_to_RGB(
        colour.xyY_to_XYZ(list(REFERENCE_COLOUR_CHECKER.data.values())),
        REFERENCE_COLOUR_CHECKER.illuminant, D65,
        colour.RGB_COLOURSPACES['sRGB'].matrix_XYZ_to_RGB)
    print(REFERENCE_SWATCHES)

    for i, swatches in enumerate(swatch_values):
        swatches_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
            swatches, D65, D65, colour.RGB_COLOURSPACES['sRGB'].matrix_RGB_to_XYZ))
    
        swatches_f = colour.colour_correction(swatches, swatches, REFERENCE_SWATCHES)
        swatches_f_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
            swatches_f, D65, D65, colour.RGB_COLOURSPACES['sRGB'].matrix_RGB_to_XYZ))


        colour_corrected_image = colour.cctf_encoding(
            colour.colour_correction(
                img1, swatches, REFERENCE_SWATCHES))
        cv2.imwrite('./assets/output/colour_corrected_image_1.jpg', colour_corrected_image)

        colour.plotting.plot_image(colour_corrected_image)

**Reply by thomasmansencal** [reply]
Hi @MadhuSN,

Thanks the image and the code were helpful. I note that you colour correct the image multiple times in the same loop, is this intended?   Ideally, you would have a single call to `colour.colour_correction`. You are also using `colour.io.read` for reading and `cv2.imwrite` for writing, I would recommend sticking to colour or cv2 definitions but not mix them.

Given a test image path and some test swatch colours, the workflow should be as follows:

1. Read the image
2. Linearise/decode the image
3. Convert the test swatch colours and reference swatch colours to the same colourspace.
     - Note 1: This can be done without a loop.
     - Note 2: The colourspace might be something else than CIE xyY.
4. Colour correct the image
5. Encode the image
6. Write the image

Cheers,

Thomas

**Reply by MadhuSN** [reply]
Hi Thomas,

I agree with you. Most of the code here is just trail and error on the reference image.

However, the images you see here are the actual image (image 1) and the image after linearising or decoding it (image 2). I can see significant reduction in the image quality here. When I encode it back it does not restore image quality. 

Please refer to the attached image. 
![op](https://user-images.githubusercontent.com/51255741/233125581-7832b477-ccc7-479d-8591-428b573bca5d.png)

This image is the final colour corrected output.

**Reply by thomasmansencal** [reply]
Do you have the measured ColorChecker values?

**Reply by tjdcs** [reply]
Are you scaling the image? It looks to me like scaling is messing with the regular interval of the lines on the image

**Reply by MadhuSN** [reply]
![swatches](https://user-images.githubusercontent.com/51255741/233591996-acae9426-81c1-4a8b-a3a2-b60f76f94a6d.png)
> Do you have the measured ColorChecker values?

Swatches [(array([ 0.15769309,  0.39624593,  0.22701213], dtype=float32), array([ 0.22282112,  0.16478506,  0.29807267], dtype=float32), array([ 0.05401623,  0.0879925 ,  0.00954172], dtype=float32), array([ 0.07344554,  0.12108786,  0.21063492], dtype=float32), array([ 0.4383969 ,  0.22097264,  0.09905463], dtype=float32), array([ 0.09021939,  0.02745266,  0.01039085], dtype=float32), array([ 0.52391714,  0.14460321,  0.00567051], dtype=float32), array([ 0.04223775,  0.05309333,  0.2878889 ], dtype=float32), array([ 0.5799691 ,  0.08710817,  0.04322479], dtype=float32), array([ 0.10819899,  0.02226758,  0.08288278], dtype=float32), array([ 0.36828655,  0.4591894 ,  0.0040894 ], dtype=float32), array([ 0.5491122,  0.2307647,  0.0061786], dtype=float32), array([ 0.02189209,  0.15521337,  0.26983383], dtype=float32), array([ 0.53995943,  0.07772446,  0.21374913], dtype=float32), array([ 0.7366904 ,  0.4929144 ,  0.00546476], dtype=float32), array([ 0.52082956,  0.0307001 ,  0.0116053 ], dtype=float32), array([ 0.05664061,  0.27818367,  0.02142191], dtype=float32), array([ 0.008285  ,  0.00946014,  0.16152452], dtype=float32), array([ 0.80920285,  0.7283255 ,  0.59500945], dtype=float32), array([ 0.54571736,  0.49706468,  0.39122885], dtype=float32), array([ 0.37800524,  0.34004018,  0.24603282], dtype=float32), array([ 0.186911  ,  0.16779776,  0.10665054], dtype=float32), array([ 0.056522  ,  0.0516582 ,  0.03113838], dtype=float32), array([ 0.01381   ,  0.01308794,  0.00981142], dtype=float32))]

These are the measured values

**Reply by MadhuSN** [reply]
> Are you scaling the image? It looks to me like scaling is messing with the regular interval of the lines on the image

I am not explicitly scaling the image. Could it be that image is being scaled else where during colour correction or while reading the image ? Currenly using "imageio" for reading and for displaying image output (i think that's the default while using colour.io)

**Reply by thomasmansencal** [reply]
There is no scaling occurring during any of those two steps!

**Reply by thomasmansencal** [reply]
Hello,

I looked at your values, and they are seemingly flipped for the first 3 rows:

```python
import colour
import numpy as np
name, data, illuminant, row, columns = colour.CCS_COLOURCHECKERS["ColorChecker24 - After November 2014"]
RGB_r = colour.XYZ_to_RGB(colour.xyY_to_XYZ(data.values()), colour.RGB_COLOURSPACES["sRGB"], illuminant)
RGB_t = np.array(([ 0.15769309, 0.39624593, 0.22701213], [ 0.22282112, 0.16478506, 0.29807267], [ 0.05401623, 0.0879925 , 0.00954172], [ 0.07344554, 0.12108786, 0.21063492], [ 0.4383969 , 0.22097264, 0.09905463], [ 0.09021939, 0.02745266, 0.01039085], [ 0.52391714, 0.14460321, 0.00567051], [ 0.04223775, 0.05309333, 0.2878889 ], [ 0.5799691 , 0.08710817, 0.04322479], [ 0.10819899, 0.02226758, 0.08288278], [ 0.36828655, 0.4591894 , 0.0040894 ], [ 0.5491122, 0.2307647, 0.0061786], [ 0.02189209, 0.15521337, 0.26983383], [ 0.53995943, 0.07772446, 0.21374913], [ 0.7366904 , 0.4929144 , 0.00546476], [ 0.52082956, 0.0307001 , 0.0116053 ], [ 0.05664061, 0.27818367, 0.02142191], [ 0.008285 , 0.00946014, 0.16152452], [ 0.80920285, 0.7283255 , 0.59500945], [ 0.54571736, 0.49706468, 0.39122885], [ 0.37800524, 0.34004018, 0.24603282], [ 0.186911 , 0.16779776, 0.10665054], [ 0.056522 , 0.0516582 , 0.03113838], [ 0.01381 , 0.01308794, 0.00981142]))
colour.plotting.plot_multi_colour_swatches(colour.cctf_encoding(np.stack((RGB_r, RGB_t), axis=1).reshape(-1, 3)), columns=6, direction="-y", compare_swatches="Stacked")
```

![image](https://user-images.githubusercontent.com/99779/233812418-5f826f67-13f5-4885-85ea-bc2257e04057.png)

**Reply by thomasmansencal** [reply]
Are you using a SpyderCHECKR24 by any chance (seems like so)?

![image](https://user-images.githubusercontent.com/99779/233886446-7e6c9183-49f0-4b3a-8d30-7fd79fa8ac13.png)

**Reply by MadhuSN** [reply]
> Are you using a SpyderCHECKR24 by any chance (seems like so)?
> 
> ![image](https://user-images.githubusercontent.com/99779/233886446-7e6c9183-49f0-4b3a-8d30-7fd79fa8ac13.png)

Hi, Yes I am using Spyder Checker 24

**Reply by MadhuSN** [reply]
I added a separate configuration for SpyderCHECKR24 at colour.CCS_COLOURCHECKERS.

**Comment by MadhuSN** [comment]
Hi, Could someone help me understand the mechanism behind the colour correction module?
 I am having some difficulty in understanding it through the code.

**Reply by thomasmansencal** [reply]
Hi,

The idea is to use a set of test values and reference values to produce a colour correction transformation through least square mapping. The colour correction can be either linear (default) or a polynomial. Having played a bit with your data a polynomial is better, I think the issue is that your imagery has some sort of S-Curve / Contrast curve applied so it cannot be linearised simply with a typical transfer function, i.e. gamma correction.

**Baseline**
![image](https://user-images.githubusercontent.com/99779/235866696-2e079dd0-f160-45a8-b920-930c01aa36ae.png)

**Linear Colour Correction (LCC)**
![image](https://user-images.githubusercontent.com/99779/235866976-38f521cf-008d-400a-9593-59f3037d7a78.png)

**Root-Polynomial Colour Correction (RPCC) - Degree 2**
![image](https://user-images.githubusercontent.com/99779/235867373-ea852a45-9b60-4ebb-bc65-e10874b896f8.png)

![image](https://user-images.githubusercontent.com/99779/235867765-8f34e24a-2d2d-4886-b945-87ca0d0f0acf.png)

While it might be tempting to increase the degree and obtain a better fitting:

![image](https://user-images.githubusercontent.com/99779/235867937-2b662bfd-7e7d-4f0b-8703-d0a0c452f2cd.png)

Because the chart has a very low number of values, anything outside is subject to typical polynomial explosion:

![image](https://user-images.githubusercontent.com/99779/235868148-b6449e16-7cc5-416f-8616-65f6ca0a4a8e.png)

```python
import colour
import numpy as np
import pandas as pd
from io import StringIO

DATA_SPYDERCHECKR_48 = StringIO(
    """
    Patch	Name	L	A	B	Rs	Gs	Bs
    1A	Card White	96.04	2.16	2.60	249	242	238
    2A	20% Gray	80.44	1.17	2.05	202	198	195
    3A	40% Gray	65.52	0.69	1.86	161	157	154
    4A	60% Gray	49.62	0.58	1.56	122	118	116
    5A	80% Gray	33.55	0.35	1.40	80	80	78
    6A	Card Black	16.91	1.43	-0.81	43	41	43
    1B	Primary Cyan	47.12	-32.52	-28.75	0	127	159
    2B	Primary Magenta	50.49	53.45	-13.55	192	75	145
    3B	Primary Yellow	83.61	3.36	87.02	245	205	0
    4B	Primary Red	41.05	60.75	31.17	186	26	51
    5B	Primary Green	54.14	-40.76	34.75	57	146	64
    6B	Primary Blue	24.75	13.78	-49.48	25	55	135
    1C	Primary Orange	60.94	38.21	61.31	222	118	32
    2C	Blueprint	37.80	7.30	-43.04	58	88	159
    3C	Pink	49.81	48.50	15.76	195	79	95
    4C	Violet	28.88	19.36	-24.48	83	58	106
    5C	Apple Green	72.45	-23.57	60.47	157	188	54
    6C	Sunflower	71.65	23.74	72.28	238	158	25
    1C	Aqua	70.19	-31.85	1.98	98	187	166
    2D	Lavender	54.38	8.84	-25.71	126	125	174
    3D	Evergreen	42.03	-15.78	22.93	82	106	60
    4D	Steel Blue	48.82	-5.11	-23.08	87	120	155
    5D	Classic Light Skin	65.10	18.14	18.68	197	145	125
    6D	Classic Dark Skin	36.13	14.15	15.78	112	76	60
    """
)
DATAFRAME_SPYDERCHECKR_48 = pd.read_csv(DATA_SPYDERCHECKR_48, sep="\t")

SPYDERCHECKR_48_sRGB = colour.utilities.tstack(
    [
        DATAFRAME_SPYDERCHECKR_48["Rs"],
        DATAFRAME_SPYDERCHECKR_48["Gs"],
        DATAFRAME_SPYDERCHECKR_48["Bs"],
    ]
)

SPYDERCHECKR_48_sRGB /= 255

# colour.plotting.plot_multi_colour_swatches(SPYDERCHECKR_48_sRGB, columns=6)

RGB_r = colour.utilities.orient(
    colour.cctf_decoding(SPYDERCHECKR_48_sRGB).reshape(4, 6, 3), "Flop"
).reshape(-1, 3)

RGB_t = np.array(
    (
        [0.15769309, 0.39624593, 0.22701213],
        [0.22282112, 0.16478506, 0.29807267],
        [0.05401623, 0.0879925, 0.00954172],
        [0.07344554, 0.12108786, 0.21063492],
        [0.4383969, 0.22097264, 0.09905463],
        [0.09021939, 0.02745266, 0.01039085],
        [0.52391714, 0.14460321, 0.00567051],
        [0.04223775, 0.05309333, 0.2878889],
        [0.5799691, 0.08710817, 0.04322479],
        [0.10819899, 0.02226758, 0.08288278],
        [0.36828655, 0.4591894, 0.0040894],
        [0.5491122, 0.2307647, 0.0061786],
        [0.02189209, 0.15521337, 0.26983383],
        [0.53995943, 0.07772446, 0.21374913],
        [0.7366904, 0.4929144, 0.00546476],
        [0.52082956, 0.0307001, 0.0116053],
        [0.05664061, 0.27818367, 0.02142191],
        [0.008285, 0.00946014, 0.16152452],
        [0.80920285, 0.7283255, 0.59500945],
        [0.54571736, 0.49706468, 0.39122885],
        [0.37800524, 0.34004018, 0.24603282],
        [0.186911, 0.16779776, 0.10665054],
        [0.056522, 0.0516582, 0.03113838],
        [0.01381, 0.01308794, 0.00981142],
    )
)
colour.plotting.plot_multi_colour_swatches(
    colour.cctf_encoding(np.stack((RGB_r, RGB_t), axis=1).reshape(-1, 3)),
    columns=6,
    direction="-y",
    compare_swatches="Stacked",
)

colour.plotting.plot_multi_colour_swatches(
    colour.cctf_encoding(
        np.stack(
            (
                RGB_r,
                np.clip(
                    colour.colour_correction(
                        RGB_t,
                        RGB_t,
                        RGB_r,
                        method="Finlayson 2015",
                        degree=2,
                        root_polynomial_expansion=True,
                    ),
                    0,
                    1,
                ),
            ),
            axis=1,
        ).reshape(-1, 3)
    ),
    columns=6,
    direction="-y",
    compare_swatches="Stacked",
)

RGB = colour.cctf_decoding(
    colour.read_image(
        "/Users/kelsolaar/Downloads/232350472-00c3ce1d-16fa-4e94-b7ce-12fdb5b71910.jpg"
    )
)

colour.write_image(
    colour.cctf_encoding(
        np.clip(
            colour.colour_correction(
                RGB,
                RGB_t,
                RGB_r,
                method="Finlayson 2015",
                degree=2,
                root_polynomial_expansion=True,
            ),
            0,
            1,
        )
    ),
    "/Users/kelsolaar/Downloads/232350472-00c3ce1d-16fa-4e94-b7ce-12fdb5b71910.png",
)
```

**Reply by MadhuSN** [reply]
Thank you very much for this. This helps a lot

Also I have one other quick doubt. Would I be able to process the same colour correction for raw image files ?

**Reply by thomasmansencal** [reply]
I would not only be possible but you would yield better results because the images should be in a scene-referred state and thus linear with respect to captured scene irradiance.

**Reply by MadhuSN** [reply]
Hi ! I have been getting this error off late 
**"AttributeError: module 'colour' has no attribute 'utilities'"**

I can see a utilities folder present in the package though.

**Reply by MadhuSN** [reply]
and also I have been trying to read raw image files using the colour library but I am unable to do so. Is there way you could help me on reading the RAW image files

Apologies I am very new to this whole image processing space. Any help would be much appreciated

**Reply by thomasmansencal** [reply]
could you do a `import colour;print(colour.__path__)` and verify that it is really `colour` at that path? There is a package on `Pypi` that has the same namespace, i.e. `colour`. Our Pypi package is `colour-science` and the namespace is `colour`:

- https://pypi.org/project/colour/ --> `colour`
- https://pypi.org/project/colour-science/ --> `colour`

As for the RAW files, with the OpenImageIO backend this might work, I actually never tried as I process them using `Rawpy`.
